pkg/responsemeta/responsemeta.go

```diff
@@ -0,0 +1,111 @@
+// metadata into the trailer of an API request.
+func NewHandler(ctx context.Context) *ResponseMetadataHandler {
+	return &ResponseMetadataHandler{
+		ctx:    ctx,
+		wasSet: false,
+	}
+}
+
+// ResponseMetadataHandler is a helper struct for tracking whether response
+// metadata was written to the trailer.
+type ResponseMetadataHandler struct {
+	ctx    context.Context
+	wasSet bool
+}
+
+// SetFor sets the response metadata in the trailer to that contained in the given
+// dispatch metadata.
+func (rmh *ResponseMetadataHandler) SetFor(dispatchMetadata *dispatch.ResponseMeta) {
+	err := setResponseMetadata(rmh.ctx, dispatchMetadata)
+	if err != nil {
+		log.Error().Err(err).Msg("error setting response metadata")
+		return
+	}
+	rmh.wasSet = true
+}
+
+// SetDefaultsIfNotCalled sets default response metadata if and only if SetFor was
+// not already called.
+func (rmh *ResponseMetadataHandler) SetDefaultsIfNotCalled() {
+	if !rmh.wasSet {
+		err := SetDefaultResponseMetadata(rmh.ctx)
+		if err != nil {
+			log.Error().Err(err).Msg("error setting default response metadata")
+			return
+		}
+		rmh.wasSet = true
+	}
+}
+
+// ResponseMetadataKey defines a key in the response metadata trailer.
+type ResponseMetadataKey string
+
+const (
+	// DispatchedOperationsCount is the key in the response trailer metadata for
+	// the number of dispatched operations that were needed to perform the overall
+	// API call.
+	DispatchedOperationsCount ResponseMetadataKey = "io.spicedb.respmeta.dispatchedoperationscount"
+
+	// CachedOperationsCount is the key in the response trailer metadata for
+	// the number of *cached* operations that would have been otherwise dispatched
+	// to perform the overall API call.
+	CachedOperationsCount ResponseMetadataKey = "io.spicedb.respmeta.cachedoperationscount"
+)
+
+// setResponseMetadata sets the external response metadata, extracted from the dispatch
+// response metadata, on the given context.
+func setResponseMetadata(ctx context.Context, dispatchMetadata *dispatch.ResponseMeta) error {
+	trailer := metadata.Pairs(
+		// DispatchedOperationsCount
+		string(DispatchedOperationsCount),
+		fmt.Sprintf("%d", dispatchMetadata.DispatchCount),
+
+		// CachedOperationsCount
+		string(CachedOperationsCount),
+		fmt.Sprintf("%d", dispatchMetadata.CachedDispatchCount),
+	)
+	return grpc.SetTrailer(ctx, trailer)
+}
+
+// SetDefaultResponseMetadata sets the default response metadata.
+func SetDefaultResponseMetadata(ctx context.Context) error {
+	trailer := metadata.Pairs(
+		// DispatchedOperationsCount
+		string(DispatchedOperationsCount),
+		"0",
+
+		// CachedOperationsCount
+		string(CachedOperationsCount),
+		"0",
+	)
+	return grpc.SetTrailer(ctx, trailer)
+}
+
+// GetIntResponseMetadata retrieves an integer value for the given key in the trailer
+// metadata of a SpiceDB API response.
+func GetIntResponseMetadata(trailer metadata.MD, key ResponseMetadataKey) (int, error) {
+	values := trailer.Get(string(key))
```

Comment:
If you can tell if they are set by the presence of this trailer, why the need to track whether they've been set separately through a bool? In fact, why do you need a class at all? Can't you just use the metadata associated with the context?

Reply:
We don't have access to the trailer when setting the data, only when reading in the client.